Blog: How to Monitor and Optimize Node-RED Flows Using Open Telemetry (#2470)
FlowFuse/website: src/blog/2024/08/opentelemetry-with-node-red.md

---
title: How to Monitor and Optimize Node-RED Flows Using Open Telemetry
subtitle: Integrating OpenTelemetry with Node-RED for Efficient Distributed Tracing
description: Learn to integrate Open Telemetry with Node-RED to track and optimize flow performance.
date: 2024-08-31
authors: ["sumit-shinde"]
image: 
tags:
   - posts
   - flowfuse
   - node-red
   - distributed tracing in node-red
   - open telemetry with node-red
   - optimizing the performance of node-red flows
   - how to monitor performance of node-red flows
---

> Have you ever found yourself frustrated by unexpected delays in your Node-RED flows, wondering where the bottlenecks are hiding? In the fast-paced world of IoT, even small latency issues can have a big impact on your system's performance. That's where Open Telemetry comes in. With its powerful distributed tracing capabilities, you can finally take control and get a clear view of how your flows are performing in real time.

Reviewer: Less ChatGPT please -- "In the fast-paced world of IoT, even small latency  ..."

<!--more-->

> In this guide, we’ll show you how to integrate Open Telemetry with Node-RED to monitor latency across your flows. By implementing distributed tracing, you’ll gain the ability to see exactly where delays occur, helping you optimize performance and ensure your IoT applications run efficiently.

Reviewer: ```suggestion
Integrating Open Telemetry with Node-RED allows you to monitor latency across your flows. By implementing distributed tracing, you’ll gain the ability to see exactly where delays occur, helping you optimize performance and ensure your IoT applications run efficiently.
```

{% endrenderFlow %}

> Deploy the above flow, and you might notice that the data printed on the debug panel appears with a delay. For demonstration purposes, we have intentionally added a delay in the function node that converts temperature data to Celsius. While this delay might be easy to spot in a simple flow, identifying such issues in larger flows with hundreds of nodes can be challenging and time-consuming. OpenTelemetry simplifies this task by providing detailed traces that help you pinpoint where delays or issues are occurring.

Reviewer: @sumitshinde-84 There's literally a delay node -- Why use High code here?

Once the trace opens, you will see the duration taken by each node to process and pass data. Notice the delay in the Kelvin to Celsius function, which is taking 2 seconds. By clicking on the green line corresponding to this function, you can view more detailed information about the trace.

!["Image showing the total duration taken by the flow"](./images/before.png "Image showing the total duration taken by the flow")
_Image showing the tototal duration taken by the flow_

Since the issue was identified in the function node, you can update the function to remove the delay which we have intentionally added. Replace the current function node with the following JavaScript code:

```js
// Extract the Kelvin value from the incoming message payload
let kelvin = msg.payload;

// Function to convert Kelvin to Celsius
function kelvinToCelsius(kelvin) {
    return kelvin - 273.15;
}

// Convert Kelvin to Celsius
let celsius = kelvinToCelsius(kelvin);

// Set the result in the message payload
msg.payload = celsius;

// Return the modified message object
return msg;

> ```

Reviewer: What's the low code equivalent? Can't the chnage node do this?

Author: Done! I have replaced the function node with a change node and added a delay node for the delay.
